include/nbl/asset/utils/CSPIRVIntrospector.h

```diff
@@ -13,172 +17,657 @@
+							arraySizes[i].isSpecConstant = true;
+						}
+					}
+
+					return range;
+				}
+				inline core::based_span<char> addString(const std::string_view str) // TODO: move to cpp
+				{
+					const auto range = alloc<char>(str.size()+1);
+					const auto out = range(m_memPool.data());
+					memcpy(out.data(),str.data(),str.size());
+					out.back() = '\0';
+					return range;
+				}
+				SDescriptorVarInfo<true>* addResource(const spirv_cross::Compiler& comp, const spirv_cross::Resource& r, IDescriptor::E_TYPE restype);
+				inline core::based_offset<SType<true>> addType(const size_t memberCount) // TODO: move to cpp
+				{
+					const auto memberStorage = allocOffset(SType<true>::StoragePerMember*memberCount);
+					auto retval = alloc<SType<true>>(1);
+					// allocated everything before touching any data stores, to avoid pointer invalidation
+					auto pType = retval(m_memPool.data()).data();
+					pType->memberCount = memberCount;
+					pType->memberInfoStorage = {memberStorage};
+					return {retval.byte_offset()};
+				}
+				void shaderMemBlockIntrospection(const spirv_cross::Compiler& comp, SMemoryBlock<true>* root, const spirv_cross::Resource& r);
+				void finalize(IShader::E_SHADER_STAGE stage);
+
+				//! debug
+				static void printExtents(std::ostringstream& out, const std::span<const SArrayInfo> counts);
+				static void printType(std::ostringstream& out, const SType<false>* counts, const uint32_t depth=0);
+				
+				IShader::E_SHADER_STAGE m_shaderStage = IShader::E_SHADER_STAGE::ESS_UNKNOWN;
+
+				// Parameters it was created with
+				SParams m_params;
+				//! Sorted by `id`
+				core::vector<SSpecConstant<>> m_specConstants; // TODO: maybe unordered_set?
+				//! Sorted by `location`
+				core::vector<SInputInterface> m_input;
+				std::variant<
+					core::vector<SFragmentOutputInterface>, // when `params.shader->getStage()==ESS_FRAGMENT`
+					core::vector<SOutputInterface> // otherwise
+				> m_output;
+				//!
+				SPushConstantInfo<> m_pushConstants = {};
+				//! Each vector is sorted by `binding`
+				constexpr static inline uint8_t DESCRIPTOR_SET_COUNT = 4;
+				core::vector<SDescriptorVarInfo<>> m_descriptorSetBindings[DESCRIPTOR_SET_COUNT];
+				// The idea is that we construct with based_span (so we can add `.data()` when accessing)
+				// then just convert from `SMemoryBlock<true>` to `SMemoryBlock<false>`
+				// in-place when filling in the `descriptorSetBinding` vector which we pass to ctor
+				core::vector<char> m_memPool;
+		};
+		class CPipelineIntrospectionData final : public CIntrospectionData
 		{
-			std::string entryPoint;
-			core::smart_refctd_ptr<const ICPUShader> cpuShader;
+			public:
+				struct SDescriptorInfo final : CIntrospectionData::SDescriptorInfo
+				{
+					inline bool isArray() const {return stride;}
+					inline bool isRuntimeSized() const {return isArray() && count==0;}
+
+					uint32_t count : 21 = 0;
+					uint32_t stride : 11 = 0;
```

Comment:
Ok because descriptor arrays can only be:
- 1 dimensional
- runtime array or fixed size
and non-arrays are treated as `[1]` this means we don't need `stride` at all.

However if SPIR-V modules don't need to agree whether a binding is runtime sized, we need
```cpp
uint32_t count : 31 = 0;
uint32_t runtimeSizedCompatible : = 0;
```

Basically then this means that `count` is the lower bound on the binding size (some SPIR-V Stage Module exists that uses the binding as `T[count]`) and `runtimeSizedCompatible` means that some SPIR-V Stage Module exists that uses `T[]`

As a sidenote we should probably store a `uint32_t minByteSize;` for SSBO and UBO descriptor types, which tells us how big we should make the buffer that backs the resource to keep all stages/SPIR-V Modules happy.

Remember that the `minByteSize` does not count runtime arrays toward itself, so 
```glsl
buffer Block {
  ... some other data ...
  T runtimeData[];
} myBuffers[6];
// minByteSize = offsetof(Block,runtimeData[0])
```